labs/00-what-is-tracing/README.md

```diff
@@ -0,0 +1,87 @@
+
+Logging is another form of tracing. It usually associates a timestamp with events to better understand the timing in the application. But we usually log only high level information, as disk space is a limited resource (right?)
+
+The `tracing` we're discussing here is high speed, low overhead tracing. With such tracing, the tracepoints can be present in the code at all time (linux has tons of tracepoints in its code, ready to be hooked to), they have a near-zero overhead when not tracing and a very low one one with a tracer enabled. Tracers can store hundreds of thousands events/second. Try that with printf!!
+
+- - -
+
+#### When to trace?
+
+Tracing is just another tool in the developer or sysadmin's toolbox. It is most often not the first one to use. But in some situations, in can be very useful, regardless of if the application to trace is instrumented or not (see next section).
+
+Tracing is best use to understand very complex system, or even simple ones, but the real added value comes when trying to understand complexity and all else fails.
+
+First, let's see what else is in the toolbox and when to best use each of those tools.
+
+* If you want to know what could/should be optimized in your application, where it spends more time, etc, then `profiling` is the best tool for the job. It will aggregate data of the different calls and show statistical results. The longer you profile, the more accurate the results will be.
+
+* If you want to debug your very complex algorithm, then a good `debugger` is the best approach, or even printfs!
+
+* If you want to know what happened at 8h06PM on your server farm when there was that big DDOS attack, then `log` files are the best first goto. The answer is probably there.
+
+When these tools fails though, `tracing` is there:
+
+* When `profiling` will miss that call at around 17min in the execution where there was a huge latency, it will either go unnoticed, or be lost as a statistic, without context or detail. `Tracing` can detect that latency and you can then zoom in on it and see everything that happened around that time and led to that moment.
+
+* When `debugging` your algorithm, or even printing the details, the bug never happen. The added latency of those techniques makes that race condition impossible to catch! Or that unit test that fails one in a 100 times. `Tracing`'s overhead is rarely high enough to remove the potential race condition, so if running with tracing enabled, when the problem happens, the trace may show what happened on the system at the time, or you may need to enable more events, but it will eventually show something.
+
+* When `log` files have put you on a trail, but looking at everything, you just can't put your finger on it. `Tracing` will show everything (well, almost) that is hidden from log data and you can have an overview of the whole system at a given time.
+
```

Comment:
When timing will affect the program. E.G. the server will timeout when you have a breakpoint